File: algorithms/js/MergeSort.js
'use strict';

function MergeSort(arr, lo, hi){
  if (lo == undefined) { lo = 0, hi = arr.length - 1; }
  if (lo >= hi){ return; }
  let mid = lo + Math.floor( ( hi - lo ) / 2 );
  MergeSort(arr, lo, mid);
  MergeSort(arr, mid + 1, hi);
  Merge(arr, lo, mid, hi);
};

function Merge(arr, lo, mid, hi){
  let arr1 = arr.slice(lo, mid + 1);
  let arr2 = arr.slice(mid + 1, hi + 1);
  let idx = lo;
  while (arr1.length || arr2.length){
    if (arr2[0] < arr1[0] || ! arr1.length ) {
      arr[idx++] = arr2.shift();
    } else {
      arr[idx++] = arr1.shift();
    }
  }
};

function SelectionMergeSort(arr, lo, hi){
  if (lo == undefined) { lo = 0, hi = arr.length - 1; }
  if (lo >= hi) { return; }
  let mid = lo + Math.floor((hi-lo)/2);
  SelectionMergeSort(arr, lo, mid);
  SelectionMergeSort(arr, mid+1, hi);
  SelectionMerge(arr, lo, mid, hi);
};

function SelectionMerge(arr, lo, mid, hi){
  if (arr.slice(lo, hi+1).length >= 4){
    Merge(arr, lo, mid, hi);
  } else {
    for (let i=lo; i<hi+1; i++){
      for (let j=i; j>0 && arr[j] < arr[j-1]; j--){
        let temp = arr[j];
        arr[j] = arr[j-1];
        arr[j-1] = temp;
      }
    }
  }
};

module.exports = {
  MergeSort,
  SelectionMergeSort,
};

File: algorithms/merge_sort.rb
# def merge_sort(arr, low=nil, high=nil)
#   high, low = arr.length - 1, 0 if low.nil?
#   return if low >= high
#   mid = low + ((high - low) / 2)
#   merge_sort(arr, low, mid)
#   merge_sort(arr, mid + 1, high)
#   merge(arr, low, mid, high)
# end

# def merge(arr, low, mid, high)
#   low_arr = arr[low..mid]
#   high_arr = arr[mid+1..high]
#   idx = low
#   while !low_arr.empty? || !high_arr.empty?
#     if low_arr.empty? || high_arr.any? && high_arr[0] < low_arr[0]
#       arr[idx] = high_arr.shift()
#       idx += 1 and next
#     end
#     high_arr.empty? || high_arr[0] >= low_arr[0]
#     arr[idx] = low_arr.shift()
#     idx += 1
#   end
# end

def merge_sort(arr)
  return arr if arr.length <= 1
  mid = arr.length / 2
  left = merge_sort(arr[0...mid])
  right = merge_sort(arr[mid..-1])
  merge(left, right)
end

def merge(l, r)
  result = []
  until l.empty? || r.empty?
    result << (l.first < r.first ? l.shift : r.shift)
  end
  result + l + r
end

arr = (0..1000).to_a.shuffle

puts merge_sort(arr)
